src/classes/OPP_StageMappingUtil.cls

```diff
@@ -91,10 +95,180 @@ public with sharing class OPP_StageMappingUtil {
-        FINALIZED_STATE_SET = new Set<String> {'finalized'};
+
+        // If either the committed state set or the finalized state set is null, query for everything and 
+        // reinitialize everything. 
+        if (COMMITTED_STATE_SET == null || FINALIZED_STATE_SET == null) {
+            Map<String, Set<String>> oppStateStages = new Map<String, Set<String>>();
+            List<Opportunity_Stage_To_State_Mapping__mdt> stageToStateRecords = [Select Id, Opportunity_Stage__c, Opportunity_State__c FROM Opportunity_Stage_To_State_Mapping__mdt];
+
+            for (Opportunity_Stage_To_State_Mapping__mdt mappingRecord : stageToStateRecords) {
+                if (mappingRecord.Opportunity_State__c != null) {
+
+                    String oppState = mappingRecord.Opportunity_State__c.trim().toUpperCase();
+                    String oppStage = mappingRecord.Opportunity_Stage__c.trim().toLowerCase();
+                    Set<String> stages = oppStateStages.containsKey(oppState) ? oppStateStages.get(oppState) : new Set<String>();
+                    stages.add(oppStage);
+                }
+            }
+
+            COMMITTED_STATE_SET = oppStateStages.containsKey(COMMITTED_STATE_VAL) ? oppStateStages.get(COMMITTED_STATE_VAL) : new Set<String>();
+            FINALIZED_STATE_SET = oppStateStages.containsKey(FINALIZED_STATE_VAL) ? oppStateStages.get(FINALIZED_STATE_VAL) : new Set<String>();
+        }
+    }
+
+    /*******************************************************************************************************
+    * @description Creates the finalized state record if there are no finalized records. It utilizes the 
+    * "Opportunity Stage when Fully Paid" field as the Opportunity stage for the finalized state. 
+    ********************************************************************************************************/
+    public static List<Metadata.CustomMetadata> createFinalizedOppStageToStateRecord(List<Metadata.CustomMetadata> metadataToInsert) {
+        npe01__Contacts_And_Orgs_Settings__c contactAndOrgCustomSetting = UTIL_CustomSEttingsFacade.getOrgContactsSettings();
+
+        // Only create the finalized opp stage to state record if there's no existing finalized record. 
+        Integer finalizedCount = [Select Count() From Opportunity_Stage_To_State_Mapping__mdt Where Opportunity_State__c =: FINALIZED_STATE_VAL];
+        
+        // Take the value, if there is one, and create a finalized Opp Stage to State custom metadata type record. 
+        if (finalizedCount == 0 && 
+            (contactAndOrgCustomSetting.Payments_Auto_Close_Stage_Name__c != null && 
+            contactAndOrgCustomSetting.Payments_Auto_Close_Stage_Name__c != '')) {
+            
+            String mdTypeName = UTIL_Namespace.StrTokenNSPrefix(OPP_STAGE_TO_STATE_METADATA_NAME);
+            String recordName = mdTypeName + '_' + FINALIZED_STATE_VAL + '_' + datetime.now().format('yyMMddHHmmss');
```

Comment:
Take mdTypeName out.  It should not be part of the Record Name.  Is there are reason you are not using DateTime.now().getTime() ?

As stated before, I would prefer a factory method to create these, taking stage and state as it's only parameter.

This section would then be:  

medataToInsert.add(createStageToStateMapping (contactAndOrgCustomSetting.Payments_Auto_Close_Stage_Name__c, FINALIZED_STATE_VAL));

Everything, including the fieldMapping, recordName, etc and setup of it, should be inside the createStageToStateMapping Method.

Reply:
I can change it so that the createStageToStateMapping method just takes in the stage and state. I'm using dateTime.now().format() because the api name can only have alphanumerics and underscore, dateTime.now() has dashes.

Reply:
Turns out we need the mdTypeName, its for creating the actual metadata record. This is an example of the one in GEM: 

"GEM_API_Settings.GEM_Config.md"

Reply:
dateTime.now().getTime() is the number of milliseconds since 1970.

Reply:
Oh sorry, I misread it. I thought you had meant just dateTime.now() instead of dateTIme.now().getTime(). I'll change it to getTime().

Reply:
Thanks!  And thanks for working through this stuff with me!!!